cranelift/codegen/src/dominator_tree.rs

```diff
@@ -350,31 +351,7 @@ impl DominatorTree {
     /// post-order. Split-invariant means that if a block is split in two, we get the same
     /// post-order except for the insertion of the new block header at the split point.
     fn push_successors(&mut self, func: &Function, block: Block) {
-        if let Some(inst) = func.layout.last_inst(block) {
-            match &func.dfg.insts[inst] {
-                ir::InstructionData::Jump {
-                    destination: succ, ..
-                } => self.push_if_unseen(succ.block(&func.dfg.value_lists)),
-                ir::InstructionData::Brif {
-                    blocks: [block_then, block_else],
-                    ..
-                } => {
-                    self.push_if_unseen(block_then.block(&func.dfg.value_lists));
-                    self.push_if_unseen(block_else.block(&func.dfg.value_lists));
-                }
-                ir::InstructionData::BranchTable {
-                    table: jt,
-                    destination: dest,
-                    ..
-                } => {
-                    for succ in func.stencil.dfg.jump_tables[*jt].iter() {
-                        self.push_if_unseen(*succ);
-                    }
-                    self.push_if_unseen(*dest);
```

Comment:
Using `visit_block_succs` changes the traversal order here, traversing the default block first.

Reply:
I think that is fine, but I am not familiar with this code.

Reply:
The doc comment for this function says the requirement is that "if a block is split in two, we get the same post-order except for the insertion of the new block header at the split point" (which it calls a "split-invariant post-order"). As far as I can tell, the order of visiting successors of a single instruction doesn't affect that property, as long as the order is stable if you compute the dominator tree multiple times.

Looking further at the history, I think a lot of this is remnants from the old "extended basic block" model. There's a giant comment in `compute_postorder`, which is immediately above this function, where as far as I can tell almost none of that comment applies to Cranelift today. Most of that comment was written when this was still called Cretonne.

I wouldn't be shocked if we don't care about "split-invariant post-orders" at all any more. Even if we do, I don't think this change breaks that property. But just in case I'm wrong about both of those, let's see if @cfallin wants to weigh in.

Reply:
Yeah, I don't think traversal order should matter here; and if it does, to the extent that it causes a bug, we should fix that.